pubsub/index.js

```diff
@@ -0,0 +1,268 @@
+  process.exit(1)
+}
+
+const IGNORED_EVENTS = new Map()
+if (process.env.IGNORED_EVENTS) {
+  // process.env.IGNORED_EVENTS is a JSON object of event_type:criteria, e.g.:
+  //
+  // {
+  //   "fxa_activity - access_token_checked": [
+  //     {
+  //       "event_properties": {
+  //         "oauth_client_id": "deadbeef"
+  //       }
+  //     },
+  //     {
+  //       "event_properties": {
+  //         "oauth_client_id": "baadf00d"
+  //       }
+  //     }
+  //   ],
+  //   "fxa_activity - access_token_created": [
+  //     {
+  //       "event_properties": {
+  //         "oauth_client_id": "deadbeef"
+  //       }
+  //     },
+  //     {
+  //       "event_properties": {
+  //         "oauth_client_id": "baadf00d"
+  //       }
+  //     }
+  //   ]
+  // }
+  Object.entries(JSON.parse(process.env.IGNORED_EVENTS)).forEach(([ type, criteria ]) => {
+    IGNORED_EVENTS.set(type, criteria)
+  })
+}
+
+const ENDPOINTS = {
+  HTTP_API: 'https://api.amplitude.com/httpapi',
+  IDENTIFY_API: 'https://api.amplitude.com/identify',
+}
+
+const KEYS = {
+  HTTP_API: 'event',
+  IDENTIFY_API: 'identification',
+}
+
+const IDENTIFY_VERBS = [ '$set', '$setOnce', '$add', '$append', '$unset' ]
+const IDENTIFY_VERBS_SET = new Set(IDENTIFY_VERBS)
+const MAX_EVENTS_PER_BATCH = 10
+const WORKER_COUNT = process.env.WORKER_COUNT ? parseInt(process.env.WORKER_COUNT) : 2
+const MESSAGES = new Map()
+
+main()
+  .catch(error => {
+    console.error(timestamp(), error.stack)
+    process.exit(1)
+  })
+
+async function main () {
+  const pubsub = new PubSub({
+    projectId: PUBSUB_PROJECT_ID
+  })
+
+  const [ subscription ] = await pubsub.createSubscription(PUBSUB_TOPIC_NAME, `fxa-amplitude-${uuid()}`)
+
+  const cargo = {
+    httpapi: setupCargo(ENDPOINTS.HTTP_API, KEYS.HTTP_API),
+    identify: setupCargo(ENDPOINTS.IDENTIFY_API, KEYS.IDENTIFY_API),
+  }
+
+  subscription.on('message', message => {
+    processMessage(cargo, message)
+  })
+
+  subscription.on('error', error => {
+    console.error(timestamp(), error.stack)
+  })
+
+  subscription.on('close', () => {
+    console.error(timestamp(), 'Error: subscription closed')
+    process.exit(1)
+  })
+}
+
+function timestamp () {
+  return new Date().toISOString()
+}
+
+function setupCargo (endpoint, key) {
+  const cargo = async.cargo(async payload => {
+    try {
+      await sendPayload(payload, endpoint, key)
+      clearMessages(payload, message => message.ack())
+      console.log(timestamp(), 'success, count =', payload.length)
+    } catch (error) {
+      console.error(timestamp(), endpoint, error.stack)
+      clearMessages(payload, message => message.nack())
+    }
+  }, MAX_EVENTS_PER_BATCH)
+
+  cargo.concurrency = WORKER_COUNT
+
+  return cargo
+}
+
+function processMessage (cargo, message) {
+  const { httpapi, identify } = parseMessage(message)
+
+  if (httpapi) {
+    MESSAGES.set(httpapi.insert_id, message)
```

Comment:
I was thinking that really it's fine for any individual request to Amplitude to fail, as long they're failing in small numbers.

But I can include those payloads in the `ack`/`nack` logic instead, it was just simpler not to because those payloads can't be uniquely identified unless we add an extra property to them.

Reply:
Pushed a fix for this in 82bfc48.